hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java

```diff
@@ -95,22 +122,30 @@ public int getClassId() {
 
     @Override
     public void writeData(ObjectDataOutput out) throws IOException {
-        out.writeObject(partitionKey);
+        if (out.getVersion().isLessThan(Versions.V5_2)) {
+            out.writeObject(getSinglePartitionKey());
+        } else {
+            out.writeObject(partitionKeys);
+        }
         out.writeObject(target);
     }
 
     @Override
     public void readData(ObjectDataInput in) throws IOException {
-        this.partitionKey = in.readObject();
+        if (in.getVersion().isLessThan(Versions.V5_2)) {
```

Comment:
I don't think this is also going to work. 

From the paths I see, the versions are only set to identified header in the enterprise edition and are simply skipped in the open-source implementation while reading. See https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java#L155-L158 

So, the version you will see here will always be the version of the _current codebase_, not the version that sent this data, if you are using open-source. Do you guys do anything special for such cases @ahmetmircik?